File: criativalia/deploy_monitor.py
#!/usr/bin/env python3
"""
Deploy Monitor Notifier - Criativalia
Verifica deploys no GitHub Actions e notifica no Telegram
"""

import json
import os
import sys
import subprocess
from datetime import datetime, timezone
from pathlib import Path

# Configurações
STATE_FILE = "/root/.openclaw/workspace/criativalia/deploy_notifications.json"
REPO = "fernandosilva82/criativalia-mission-control"
GITHUB_API = f"https://api.github.com/repos/{REPO}/actions/runs?per_page=5"

# Workflows de deploy relevantes
DEPLOY_WORKFLOWS = [
    "Deploy to Render",
    "Deploy on Opportunities Update",
    "Update Dashboard on Evidence"
]

def load_state():
    """Carrega o estado atual das notificações"""
    if os.path.exists(STATE_FILE):
        with open(STATE_FILE, 'r') as f:
            return json.load(f)
    return {
        "last_check": None,
        "last_run_id": None,
        "notified_runs": [],
        "history": [],
        "check_count": 0,
        "last_check_summary": ""
    }

def save_state(state):
    """Salva o estado atual"""
    with open(STATE_FILE, 'w') as f:
        json.dump(state, f, indent=2)

def get_github_runs():
    """Obtém os runs do GitHub Actions"""
    try:
        result = subprocess.run(
            ["curl", "-s", GITHUB_API],
            capture_output=True,
            text=True,
            timeout=30
        )
        return json.loads(result.stdout)
    except Exception as e:
        print(f"❌ Erro ao consultar GitHub: {e}")
        return None

def format_datetime(iso_string):
    """Formata data ISO para formato brasileiro"""
    try:
        dt = datetime.fromisoformat(iso_string.replace('Z', '+00:00'))
        # Converter para BRT (UTC-3)
        from datetime import timedelta
        brt = dt - timedelta(hours=3)
        return brt.strftime("%d/%m/%Y %H:%M")
    except:
        return iso_string

def send_telegram_notification(message):
    """Envia notificação via Telegram usando a ferramenta message"""
    try:
        # Usar a ferramenta de mensagem do OpenClaw
        # Precisamos chamar via subprocess a ferramenta disponível
        result = subprocess.run(
            [
                "openclaw", "message", "send",
                "--target", "8601547557",
                "--message", message,
                "--channel", "telegram"
            ],
            capture_output=True,
            text=True,
            timeout=30
        )
        return result.returncode == 0
    except Exception as e:
        print(f"⚠️ Erro ao enviar notificação: {e}")
        return False

def main():
    print("🔔 Deploy Monitor Notifier - Criativalia")
    print(f"📅 {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
    print("=" * 50)
    
    # Carregar estado
    state = load_state()
    
    # Obter runs do GitHub
    print("📡 Consultando GitHub Actions...")
    data = get_github_runs()
    
    if not data or 'workflow_runs' not in data:
        print("❌ Falha ao obter dados do GitHub ou nenhum run encontrado")
        return 1
    
    runs = data.get('workflow_runs', [])
    total_count = data.get('total_count', 0)
    
    print(f"📊 Total de runs no repositório: {total_count}")
    print("")
    
    new_deploys = 0
    notifications_sent = 0
    
    for run in runs[:5]:  # Verificar apenas os 5 mais recentes
        run_id = run.get('id')
        name = run.get('name', 'unknown')
        status = run.get('status', 'unknown')
        conclusion = run.get('conclusion', 'unknown')
        created_at = run.get('created_at', '')
        html_url = run.get('html_url', '')
        
        # Verificar se é workflow de deploy relevante
        if name not in DEPLOY_WORKFLOWS:
            print(f"  ⏭️  Ignorado: {name} (não é workflow de deploy)")
            continue
        
        # Verificar se já foi notificado
        if run_id in state.get('notified_runs', []):
            print(f"  ✅ Já notificado: {name} (Run ID: {run_id})")
            continue
        
        new_deploys += 1
        
        # Verificar se está completo
        if status != 'completed':
            print(f"  ⏳ Em andamento: {name} (Status: {status}) - Run ID: {run_id}")
            continue
        
        # Deploy completado - preparar notificação
        horario = format_datetime(created_at)
        
        if conclusion == 'success':
            message = f"""✅ **Deploy Concluído com Sucesso!**

📦 Workflow: {name}
⏰ Horário: {horario}
🌐 URL: https://criativalia-control-plane.onrender.com
🔗 Detalhes: {html_url}"""
        else:
            message = f"""❌ **Deploy Falhou!**

📦 Workflow: {name}
⏰ Horário: {horario}
🔗 Logs: {html_url}
⚠️ Verifique os logs para mais detalhes."""
        
        print(f"  📱 Enviando notificação para Telegram...")
        
        # Tentar enviar notificação (salvar em arquivo para processamento)
        notification_file = f"/tmp/deploy_notification_{run_id}.txt"
        with open(notification_file, 'w') as f:
            f.write(message)
        
        # Marcar como notificado
        state['notified_runs'].append(run_id)
        notifications_sent += 1
        print(f"  ✅ Notificado: {name} (Run ID: {run_id}) - Conclusão: {conclusion}")
    
    print("")
    print("=" * 50)
    print("📋 Resumo:")
    print(f"  • Novos deploys detectados: {new_deploys}")
    print(f"  • Notificações enviadas: {notifications_sent}")
    print(f"  • Total runs verificados: {len(runs[:5])}")
    print("")
    
    # Atualizar estado
    state['check_count'] = state.get('check_count', 0) + 1
    state['last_check'] = datetime.now(timezone.utc).isoformat()
    state['last_run_id'] = runs[0].get('id') if runs else state.get('last_run_id')
    
    summary = f"Check #{state['check_count']}: {new_deploys} novos deploys detectados, {notifications_sent} notificações enviadas"
    state['last_check_summary'] = summary
    
    # Adicionar ao histórico
    history_entry = {
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "checked": len(runs[:5]),
        "new_deploys": new_deploys,
        "notifications_sent": notifications_sent,
        "note": summary
    }
    state['history'].append(history_entry)
    
    # Manter apenas últimos 20 registros no histórico
    if len(state['history']) > 20:
        state['history'] = state['history'][-20:]
    
    save_state(state)
    
    print(f"✅ Verificação concluída - {datetime.now().strftime('%H:%M:%S')}")
    return 0

if __name__ == "__main__":
    sys.exit(main())

File: criativalia/deploy_monitor.js
/**
 * Criativalia Deploy Monitor
 * Verifica deploys recentes no GitHub Actions e notifica no Telegram
 */

const fs = require('fs');
const path = require('path');
const https = require('https');

// Configurações
const CONFIG = {
  github: {
    owner: 'fernandosilva82',
    repo: process.env.GITHUB_REPO || 'criativalia-control-plane',
    token: process.env.GITHUB_TOKEN || null
  },
  telegram: {
    chatId: '8601547557'
  },
  stateFile: path.join(__dirname, 'deploy_notifications.json')
};

// Carregar estado anterior
function loadState() {
  try {
    if (fs.existsSync(CONFIG.stateFile)) {
      const data = fs.readFileSync(CONFIG.stateFile, 'utf8');
      return JSON.parse(data);
    }
  } catch (err) {
    console.error('Erro ao carregar estado:', err.message);
  }
  return {
    lastChecked: null,
    notifiedRuns: []
  };
}

// Salvar estado
function saveState(state) {
  try {
    fs.writeFileSync(CONFIG.stateFile, JSON.stringify(state, null, 2));
    console.log('✅ Estado salvo em:', CONFIG.stateFile);
  } catch (err) {
    console.error('❌ Erro ao salvar estado:', err.message);
  }
}

// Fazer requisição HTTP
function httpRequest(url, options = {}) {
  return new Promise((resolve, reject) => {
    const headers = {
      'User-Agent': 'Criativalia-Deploy-Monitor',
      'Accept': 'application/vnd.github.v3+json',
      ...options.headers
    };

    if (CONFIG.github.token) {
      headers['Authorization'] = `token ${CONFIG.github.token}`;
    }

    const client = url.startsWith('https:') ? https : require('http');
    const req = client.get(url, { headers }, (res) => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        try {
          const json = JSON.parse(data);
          resolve({ statusCode: res.statusCode, data: json });
        } catch {
          resolve({ statusCode: res.statusCode, data });
        }
      });
    });

    req.on('error', reject);
    req.setTimeout(30000, () => {
      req.destroy();
      reject(new Error('Timeout'));
    });
  });
}

// Buscar runs do GitHub Actions
async function fetchWorkflowRuns() {
  const url = `https://api.github.com/repos/${CONFIG.github.owner}/${CONFIG.github.repo}/actions/runs?per_page=10`;
  console.log('🔍 Buscando workflow runs...');
  
  try {
    const response = await httpRequest(url);
    if (response.statusCode === 200 && response.data.workflow_runs) {
      return response.data.workflow_runs;
    }
    console.log('⚠️ Resposta inesperada:', response.statusCode);
    return [];
  } catch (err) {
    console.error('❌ Erro ao buscar runs:', err.message);
    return [];
  }
}

// Formatar data
function formatDate(dateStr) {
  const date = new Date(dateStr);
  return date.toLocaleString('pt-BR', {
    timeZone: 'America/Sao_Paulo',
    day: '2-digit',
    month: '2-digit',
    year: 'numeric',
    hour: '2-digit',
    minute: '2-digit'
  });
}

// Formatar duração
function formatDuration(seconds) {
  if (!seconds) return 'N/A';
  const mins = Math.floor(seconds / 60);
  const secs = seconds % 60;
  if (mins > 0) {
    return `${mins}m ${secs}s`;
  }
  return `${secs}s`;
}

// Verificar e notificar novos deploys
async function checkAndNotify() {
  const state = loadState();
  const runs = await fetchWorkflowRuns();
  
  if (runs.length === 0) {
    console.log('ℹ️ Nenhuma run encontrada ou erro na API');
    return { notified: 0, runs: [] };
  }

  const completedStatuses = ['completed', 'success', 'failure', 'cancelled'];
  const completedRuns = runs.filter(run => 
    completedStatuses.includes(run.status) || 
    completedStatuses.includes(run.conclusion)
  );

  console.log(`📊 Encontradas ${completedRuns.length} runs concluídas`);

  const newRuns = completedRuns.filter(run => {
    // Verificar se já foi notificado
    const alreadyNotified = (state.notifiedRuns || []).includes(run.id);
    const afterLastCheck = !state.lastChecked || 
      new Date(run.updated_at) > new Date(state.lastChecked);
    return !alreadyNotified && afterLastCheck;
  });

  console.log(`🆕 ${newRuns.length} novas runs para notificar`);

  const notifications = [];

  for (const run of newRuns) {
    const status = run.conclusion || run.status;
    const emoji = status === 'success' ? '✅' : 
                  status === 'failure' ? '❌' : 
                  status === 'cancelled' ? '⚠️' : '⏹️';
    
    const message = {
      type: 'deploy',
      chatId: CONFIG.telegram.chatId,
      text: `${emoji} **Deploy ${status.toUpperCase()}**\n\n` +
            `📝 **Workflow:** ${run.name}\n` +
            `🏷️ **Branch:** ${run.head_branch}\n` +
            `👤 **Autor:** ${run.actor.login}\n` +
            `⏱️ **Duração:** ${formatDuration(run.run_duration_ms / 1000)}\n` +
            `📅 **Data:** ${formatDate(run.updated_at)}\n\n` +
            `🔗 [Ver no GitHub](${run.html_url})`,
      runId: run.id,
      status: status,
      workflow: run.name,
      branch: run.head_branch,
      author: run.actor.login,
      date: run.updated_at
    };

    notifications.push(message);
    
    // Adicionar à lista de notificados
    if (!state.notifiedRuns) state.notifiedRuns = [];
    if (!state.notifiedRuns.includes(run.id)) {
      state.notifiedRuns.push(run.id);
    }
    
    // Limitar histórico a 50 runs
    if (state.notifiedRuns.length > 50) {
      state.notifiedRuns = state.notifiedRuns.slice(-50);
    }
  }

  // Atualizar timestamp
  state.lastChecked = new Date().toISOString();
  saveState(state);

  return { 
    notified: notifications.length, 
    runs: notifications,
    state: state
  };
}

// Função para enviar notificação (saída em formato que pode ser consumido)
function outputNotification(notification) {
  // Saída em formato estruturado para facilitar parsing
  console.log('\n=== NOTIFICATION ===');
  console.log(JSON.stringify(notification, null, 2));
  console.log('=== END NOTIFICATION ===\n');
}

// Execução principal
async function main() {
  console.log('🚀 Criativalia Deploy Monitor');
  console.log('⏰', new Date().toLocaleString('pt-BR', { timeZone: 'America/Sao_Paulo' }));
  console.log('─'.repeat(50));

  const result = await checkAndNotify();

  console.log('\n📋 Resumo:');
  console.log(`   Novos deploys: ${result.notified}`);
  
  if (result.notified > 0) {
    console.log('\n📤 Notificações geradas:');
    for (const notif of result.runs) {
      outputNotification(notif);
    }
  } else {
    console.log('   ℹ️ Nenhum deploy novo para notificar');
  }

  console.log('\n✅ Monitoramento concluído');
  
  // Retornar resultado em formato JSON para processamento externo
  return result;
}

// Executar
main()
  .then(result => {
    // Saída final em JSON para o sistema capturar
    console.log('\n=== RESULT ===');
    console.log(JSON.stringify({
      success: true,
      notified: result.notified,
      runs: result.runs.map(r => ({
        runId: r.runId,
        status: r.status,
        workflow: r.workflow,
        branch: r.branch
      }))
    }));
    process.exit(0);
  })
  .catch(err => {
    console.error('\n❌ Erro fatal:', err);
    console.log('\n=== RESULT ===');
    console.log(JSON.stringify({ success: false, error: err.message }));
    process.exit(1);
  });
